pkg/apis/events/v1alpha1/cloudpubsubsource_validation.go

```diff
@@ -40,6 +42,7 @@ const (
 
 func (current *CloudPubSubSource) Validate(ctx context.Context) *apis.FieldError {
 	errs := current.Spec.Validate(ctx).ViaField("spec")
+	errs = duckv1alpha1.ValidateClusterNameAnnotation(current.Annotations, errs, metadataClient.NewDefaultMetadataClient())
```

Comment:
They will be updated by the user.

Together with comment https://github.com/google/knative-gcp/pull/950#discussion_r421149115
It looks like that I can remove clusterName validation, (which means allow the clusterName to be empty in defaulting webhook, even if it is on GKE). I feel that not every GCP resource names need to include clusterName like WI, blocking this from the defaulting webhook may not that user-friendly. Let the Identity reconciler check clusterName annotation again when we want it to generate KSA name(we already do that in the Identity reconciler), then print out error if user need to update annotation.

In the meanwhile, adding immutability check, once clusterName annotation is set, it could not be changed.

Do you think that is proper ?

Reply:
We talked about this over video. It seems like a better user experience to remove the validation and replace it with an error message about the missing annotation in the object status. This makes the upgrade easier since missing cluster names can be corrected gradually.